### PR TITLE
EXOGW-349 - Feature: Configurable `id` field

### PR DESCRIPTION
**Pull request by thekevinbrown**
Previously primary keys had to be called `id` on every entity.

With this PR, entities can define whatever field they like as the primary key. If no field is marked as the primary key we default to `id` still. Fields can be marked as primary key in the `@Field` decorator options.

**Limitations**
- We still do not support compound keys. Only one primary key is accepted.

Screenshot 2024-05-23 at 4 33 40 PM

### CODE REVIEW

File: src/packages/aws-cognito/src/provider/base-provider.ts [hunk 1]
```diff
@@ -1,24 +1,26 @@
-import { Filter, PaginationOptions, BackendProvider, WithId } from '@exogee/graphweaver';
+import { Filter, PaginationOptions, BackendProvider } from '@exogee/graphweaver';
 
-export interface ProviderOptions<Entity, Context, DataEntity> {
+export interface WithId {
```
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/packages/aws-cognito/src/provider/base-provider.ts:3]: Yeah, I moved it into the cognito package specifically so as not to have to refactor how that works for now.

File: src/packages/aws-cognito/src/entities/graphql-entity.ts [hunk 2]
```diff
@@ -1,9 +1,11 @@
-import { GraphQLEntity, ID, Entity, Field } from '@exogee/graphweaver';
+import { ID, Entity, Field } from '@exogee/graphweaver';
 
 import { CognitoUserBackendEntity } from './backend-entity';
 
 @Entity('CognitoUser')
-export class CognitoUser extends GraphQLEntity<CognitoUserBackendEntity> {
+export class CognitoUser {
+	public dataEntity!: CognitoUserBackendEntity;
```
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/packages/aws-cognito/src/entities/graphql-entity.ts:7]: Indeed!

File: src/packages/auth/src/authentication/apollo/plugin.ts [hunk 3]
```diff
@@ -98,12 +98,12 @@ export const applyDefaultMetadataACL = () => {
 	}
 };
 
-type AuthApolloPluginOptions<D, R> = {
+type AuthApolloPluginOptions<D extends ApiKeyEntity<R>, R> = {
 	implicitAllow?: boolean;
-	apiKeyDataProvider?: BackendProvider<D, ApiKeyStorage<R>>;
+	apiKeyDataProvider?: BackendProvider<D>;
```
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/packages/auth/src/authentication/apollo/plugin.ts]: Done

File: src/packages/admin-ui-components/src/combo-box/component.tsx [hunk 4]
```diff
@@ -84,7 +84,10 @@ export const ComboBox = ({
 	};
 
 	// Store the selected ids in an array for easy lookup
-	const selectedIds = useMemo(() => selectedItems.map((item) => item.value), [selectedItems]);
+	const selectedIds = useMemo(
+		() => new Set(selectedItems.map((item) => item.value)),
```
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/packages/admin-ui-components/src/combo-box/component.tsx:88]: It's the primary key for the entity, so if two had the same value it'd be problematic.

File: src/packages/admin-ui-components/src/detail-panel/util.test.ts [hunk 5]
```diff
@@ -2,109 +2,107 @@ import { describe, it, expect } from 'vitest';
...
-				},
-				{
-					id: '14',
-				},
-			],
-		};
-
-		expect(mapFormikValuesToGqlRequestValues(formValues)).toEqual(expected);
-	});
+	// it('Should replace objects representing related entities with an object containing just an ID field', () => {
```
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/packages/admin-ui-components/src/detail-panel/util.test.ts:5]: Yup, will do.

File: .github/workflows/unit-tests.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
+name: Unit Tests
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+
+jobs:
+  auth-end-to-end-tests:
```
Comment (steve-keep) [inline .github/workflows/unit-tests.yml:8]: `unit-tests`

File: src/examples/sqlite/src/backend/schema/customer.ts [hunk 7]
```diff
@@ -8,11 +8,11 @@ import { connection } from '../database';
 @Entity('Customer', {
 	provider: new MikroBackendProvider(OrmCustomer, connection),
 })
-export class Customer extends GraphQLEntity<OrmCustomer> {
+export class Customer {
 	public dataEntity!: OrmCustomer;
```
Comment (steve-keep) [inline src/examples/sqlite/src/backend/schema/customer.ts:12]: This can now go.

File: src/examples/sqlite/src/backend/schema/invoice-line.ts [hunk 8]
```diff
@@ -8,16 +8,18 @@ import { connection } from '../database';
 @Entity('InvoiceLine', {
 	provider: new MikroBackendProvider(OrmInvoiceLine, connection),
 })
-export class InvoiceLine extends GraphQLEntity<OrmInvoiceLine> {
+export class InvoiceLine {
 	public dataEntity!: OrmInvoiceLine;
```
Comment (steve-keep) [inline src/examples/sqlite/src/backend/schema/invoice-line.ts:12]: This can now go.

File: src/packages/end-to-end/src/__tests__/api/auth/acl/access-control-list.test.ts [hunk 9]
```diff
@@ -21,12 +21,12 @@ const user = new UserProfile({
...
 
-const artistDataProvider = new BaseDataProvider<any, Artist>('artist');
+const artistDataProvider = new BaseDataProvider<any>('artist');
 
 @Entity('Artist', {
 	provider: artistDataProvider,
 })
-export class Artist extends GraphQLEntity<any> {
+export class Artist {
 	public dataEntity!: any;
```
Comment (steve-keep) [inline src/packages/end-to-end/src/__tests__/api/auth/acl/access-control-list.test.ts:30]: Ok, there are a few so we can just search for any `public dataEntity!` and remove.

File: src/packages/admin-ui-components/src/detail-panel/component.tsx [hunk 10]
```diff
@@ -346,18 +346,20 @@ export const DetailPanel = () => {
 
 	const handleOnSubmit = async (formValues: any, actions: FormikHelpers<any>) => {
 		// Format form values as GraphQL input parameters
-		const values = mapFormikValuesToGqlRequestValues(formValues);
+		const values = mapFormikValuesToGqlRequestValues(selectedEntity, entityByName, formValues);
+
+		console.log(formValues, values);
```
Comment (steve-keep) [inline src/packages/admin-ui-components/src/detail-panel/component.tsx:351]:
```suggestion
```

File: src/packages/apollo-client/package.json [hunk 11]
```diff
@@ -21,7 +21,11 @@
 	"files": [
 		"lib"
 	],
+	"peerDependencies": {
+		"@apollo/client": "^3.10.4"
```
Comment (steve-keep) [inline src/packages/apollo-client/package.json:25]:
```suggestion
		"@apollo/client": "3.10.4"
```

File: src/packages/auth/src/authentication/token/auth-token.ts [hunk 12]
```diff
@@ -110,9 +110,8 @@ export class AuthTokenProvider implements BaseAuthTokenProvider {
 	}
 }
 
-export const verifyAndCreateTokenFromAuthToken = (authToken: AuthToken) => {
+export const verifyToken = (authToken?: AuthToken) => {
```
Comment (steve-keep) [inline src/packages/auth/src/authentication/token/auth-token.ts:113]: Should we remove this function now as its only checking if undefined. That check could happen in the `stepUpToken` and `generateToken` functions on `tokenProvider`

File: src/packages/end-to-end/src/types.ts [hunk 13]
```diff
@@ -75,7 +75,7 @@ export type AdminUiMetadata = {
 export type Album = {
```
Comment (steve-keep) [inline src/packages/end-to-end/src/types.ts:75]: Why is this file here?

File: src/packages/mikroorm/src/utils/change-tracker.ts [hunk 14]
```diff
@@ -67,33 +67,35 @@ const dataForChangeSet = <T extends TrackedEntity<T>>(cs: ChangeSet<TrackedEntit
...
-				? (from as Reference<AnyEntity>).getProperty('id')
+				? (from as Reference<AnyEntity>).getProperty(primaryKey)
 				: from;
 			from = {
 				reference: {
 					type: relatedToEntity.constructor.name,
 					id,
 				},
-			};
+			} as any;
```
Comment (steve-keep) [inline src/packages/mikroorm/src/utils/change-tracker.ts:91]: Can these be typed line 91 and 98?

**Review by steve-keep — APPROVED** [review]
🎉 Nothing major or blocking but a few things we could tidy up.

File: src/packages/auth/src/authentication/token/auth-token.ts (hunk 12 above)
Comment (thekevinbrown) [reply src/packages/auth/src/authentication/token/auth-token.ts]: Done in a9ce9b9b3a5235265e10d18fcbcb2c2dd9119f6f
